design/one-pager-composition-revisions.md

```diff
@@ -0,0 +1,300 @@
+  the new `Composition` is the same person who will update the XRs that consume
+  it.
+
+It must be possible to introduce this functionality in a measured, backward
+compatible way. Crossplane's behaviour and v1 APIs should not change for anyone
+who does not opt into this new functionality.
+
+## Proposal
+
+This document proposes the introduction of a new type - `CompositionRevision`.
+XRs will use a `CompositionRevision`, not a `Composition` to determine which
+managed resources should satisfy an XR.
+
+Platform teams will still create and update `Composition` resources - the new
+`CompositionRevision` resources will be created automatically by Crossplane. A
+controller will create an immutable `CompositionRevision` corresponding to each
+update to (or 'revision' of) a `Composition`. This allows an XR to be 'pinned'
+to a particular revision of a `Composition`, thus allowing the `Composition` to
+be updated without inherently affecting the XRs that (indirectly) use it. The
+new `CompositionRevision` resource's schema will be a superset of the existing
+`Composition` schema.
+
+Each time a `Composition` is updated a controller will automatically create a
+`CompositionRevision` - an immutable snapshot of that particular 'revision' of
+the `Composition`. The `CompositionRevision` schema will be a superset of the
+`Composition` schema.
+
+![XR to CompositionRevision relationship][xr-to-revision]
+
+For example the `Composition` below would result in the creation of the
+subsequent `CompositionRevision`:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: CompositePostgreSQLInstance
+  writeConnectionSecretsToNamespace: crossplane-system
+  resources:
+    - name: cloudsqlinstance
+      base:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
+        spec:
+          forProvider:
+            databaseVersion: POSTGRES_9_6
+            region: us-central1
+            settings:
+              tier: db-custom-1-3840
+              dataDiskType: PD_SSD
+      patches:
+        - fromFieldPath: "spec.parameters.storageGB"
+          toFieldPath: "spec.forProvider.settings.dataDiskSizeGb"
+```
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: CompositionRevision
+metadata:
+  # The revision's name is derived from the composition's name per the API
+  # server's metadata.generateName support.
+  name: example-jfjn3
+  ownerReferences:
+  # The Composition is the controller reference of the revision. When a
+  # Composition is deleted all of its CompositionRevisions are garbage
+  # collected.
+  - apiVersion: apiextensions.crossplane.io/v1
+    kind: Composition
+    name: example
+    controller: true
+spec:
+  # Each revision includes an integer revision number. This number increases
+  # monotonically as the associated composition is updated.
+  revision: 1
```

Comment:
Can we have this revision number in the name as well for easier browsing? i.e. `example-1-jfjn3` instead of `example-jfjn3`

Reply:
That sounds like a good approach. Honestly we could probably avoid the random suffix if we just used the revision number. 🤔

Reply:
Updated the doc to reflect that names will just be `compname-revnum`, like `example-1`.